Stash #2121

Shirkit opened the pull request:
First, thanks for this great mod! I loved playing with it, so I wanted to give back a little!

Implements #2121 

# Changes proposed in this pull request:
- Added a Stash, that will have items inside that the DeliveryMan will pickup and move to the Warehouse.

So, this **incomplete** PR is an attempt to implement #2121 , although just the first comment, as the second one I felt it was beyond the scope of the Stash, as I thought this should be kept to a minimal functionality.

## Core Functionality

The Core functionality of how I thought was this: when an item gets deposited, it bumps the priority up to 10, to encourage the DeliveryMan to pickup items ASAP on them. When the DeliveryMan picks up the items, it drops the priority to a minimum. So the bumping and the dropping are being handled by the TileEntity and the Inventory Handler.

## Found issues

I noticed that the dropping of the priority should be done automatically, but it's only done when it's the last building on the list? I didn't quite understand why this didn't work: https://github.com/ldtteam/minecolonies/blob/1fdde5f3da396a306134f6cffb814e17bb79d1f0/src/main/java/com/minecolonies/coremod/entity/ai/citizen/deliveryman/EntityAIWorkDeliveryman.java#L217

## Temporary work

For now, I don't have a fixed GUI. Suggestions are welcome for functionality.
Also, I don't have a model. If you guys have a suggestion, one to spare or a software for me to take a look (I will try Blockbench).
Also, I don't know how you guys manage your language files, so, for now, I still have no lang strings for it.

## Conclusion

I'll update after I have done more work, I just wanted to say the work is being done, and get feedback if you think the direction is right. For now, only the first commit actually needs to be looked at https://github.com/ldtteam/minecolonies/commit/135da011e962de18864533903ca9bff9b1d92186

Review and suggestions are welcome

Shirkit committed:
Stash is added

This commit contains the logic behind the worker hut, the block and the TileEntity.
I've created a TileEntity as I've found no pratical way to change the `inventory` from `AbstractTileEntityRack` for all instances of Stash. This obviously can be changed if a better way comes up. For now, it's a protected field and has no way to be re-assigned, so this is what I came up.

Shirkit committed:
GUI Work is totally temporary

The GUI is totally temporary for now, I just copied from other buildings. Also, I only managed to set the correct values when opened by using the `onOpened` event method, although it's not used anywhere else.

Shirkit committed:
Texture and Model are temporary

I've just copied the PostBox for now, as I don't have a model for it yet.

Shirkit committed:
Recipe is temporary as well

I thought of making it a double chest worth actually, but prolly 27 slots fits better the style.

Shirkit committed:
Changes to existing files to support the Stash

Registering the Block, as well as the Stash Building, the GUI element and the TileEntity, as well as allowing to change priority on the Network messages.
Also, no lang files were edited, and the GUI needs a proper translation for Stash

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/ldtteam/minecolonies?pullRequest=4545) <br/>All committers have signed the CLA.

reggaemuffin commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/Stash.java  1
- src/main/java/com/minecolonies/coremod/network/messages/ChangeDeliveryPriorityMessage.java  1
- src/api/java/com/minecolonies/api/tileentities/TileEntityStash.java  6
- src/main/java/com/minecolonies/coremod/client/gui/WindowStash.java  3
- src/main/java/com/minecolonies/coremod/network/messages/ChangeDeliveryPriorityStateMessage.java  1
         

Clones added
============
- src/main/java/com/minecolonies/coremod/blocks/huts/BlockStash.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Minecolonies/minecolonies/pullRequest?prid=5282331&bid=17293574)

Shirkit committed:
Typo on my part

Was just testing stuff

reggaemuffin commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/main/java/com/minecolonies/coremod/blocks/huts/BlockStash.java  4
- src/main/java/com/minecolonies/coremod/network/messages/ChangeDeliveryPriorityMessage.java  1
- src/main/java/com/minecolonies/coremod/client/gui/WindowStash.java  3
- src/main/java/com/minecolonies/coremod/network/messages/ChangeDeliveryPriorityStateMessage.java  1
- src/api/java/com/minecolonies/api/tileentities/TileEntityStash.java  6
- src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/Stash.java  1
         

Clones added
============
- src/main/java/com/minecolonies/coremod/blocks/huts/BlockStash.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Minecolonies/minecolonies/pullRequest?prid=5282331&bid=17293574)

Raycoms committed:
Merge branch 'version/1.15' into version/1.15

Raycoms commented on src/api/java/com/minecolonies/api/tileentities/TileEntityStash.java:
```diff
@@ -0,0 +1,61 @@
+        IBuildingContainer building = getBuilding();
+        if (building != null && !building.isPriorityStatic()) {
+            if (isEmpty)
+                building.alterPickUpPriority(MAX_PRIO*-1);
```
this would have to be 0. The lowest prio is 0 atm. That's probably why it doesn't work?

Raycoms commented on src/api/java/com/minecolonies/api/tileentities/TileEntityStash.java:
```diff
@@ -0,0 +1,61 @@
+        inventory = new NotifyingRackInventory(DEFAULT_SIZE);
+    }
+
+    private void buildingInventoryChanged(boolean isEmpty) {
```
Can you format those to our code style and add javadoc?

reggaemuffin commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/main/java/com/minecolonies/coremod/network/messages/ChangeDeliveryPriorityMessage.java  1
- src/main/java/com/minecolonies/coremod/blocks/huts/BlockStash.java  4
- src/api/java/com/minecolonies/api/tileentities/TileEntityStash.java  6
- src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/Stash.java  1
- src/main/java/com/minecolonies/coremod/client/gui/WindowStash.java  3
- src/main/java/com/minecolonies/coremod/network/messages/ChangeDeliveryPriorityStateMessage.java  1
         

Clones added
============
- src/main/java/com/minecolonies/coremod/blocks/huts/BlockStash.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Minecolonies/minecolonies/pullRequest?prid=5282331&bid=17293574)

Raycoms commented on src/main/java/com/minecolonies/coremod/client/gui/WindowStash.java:
```diff
@@ -0,0 +1,126 @@
+/**
+ * Window for the replace block GUI.
+ */
+public class WindowStash extends AbstractWindowSkeleton implements ButtonHandler
```
I would probably not have it have a GUI at all. Only the inventory GUI should be enough.

Raycoms commented:
First of all, thanks for your contribution.

I added a few small comments to it. Overall, I think the block doesn't need a GUI at all. The prio can be adapted automatically only.

The best way to "reset" the gather prio is probably in the dman "after he finished gathering".
This way you got a guaranteed reset even if he bumps it during gathering up to 10.

Finally, I believe that there are a few other places where you have to add an exception for this block so that it is not recognized as a building (Check on the instanceof BlockPosBox calls.)

Shirkit commented:
Yeah, I thought about not having a GUI, I will do that, makes more sense.

For the gathering, I'll take a closer look at the AI to ser why exactly it didn't reset. That was my plan all along, but the reset didn't seem to work, as pointed on found issues. Gonna take a closer look o that.

Definitively will add Javadoc and formatting, I just noticed the checkstylefile.

Will take a deeper look on the Building as you said. I had it originally, copying from PostBox, but I had removed for testing, and forgot to add back.

Thanks for the insanely fast feedback, I really wasn't expecting this early!

Cheers!

Shirkit committed:
New model

Derived from PostBox. I thought it would be better than to just re-use the same model

Shirkit committed:
Removed the GUI

It only uses empty files to go around the checks from Validate.notNull, thus using a dummy empty view.

Shirkit committed:
Update on block model's bounding box

Default 1 block size instead of a custom size.

Shirkit committed:
Updated the logic to let the DeliveryMan reset the priority

Shirkit committed:
Added the Stash to the ignore along with PostBox

But it's still not working when placing.

Shirkit committed:
Merge branch 'version/1.15' of https://github.com/Shirkit/minecolonies into version/1.15

reggaemuffin commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/api/java/com/minecolonies/api/tileentities/TileEntityStash.java  6
- src/main/java/com/minecolonies/coremod/network/messages/ChangeDeliveryPriorityMessage.java  1
- src/main/java/com/minecolonies/coremod/blocks/huts/BlockStash.java  5
- src/main/java/com/minecolonies/coremod/colony/buildings/views/EmptyView.java  1
- src/main/java/com/minecolonies/coremod/network/messages/ChangeDeliveryPriorityStateMessage.java  1
- src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/Stash.java  1
         

Clones added
============
- src/main/java/com/minecolonies/coremod/blocks/huts/BlockStash.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Minecolonies/minecolonies/pullRequest?prid=5282331&bid=17293574)

Shirkit committed:
Javadocs

reggaemuffin commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/main/java/com/minecolonies/coremod/network/messages/ChangeDeliveryPriorityStateMessage.java  1
- src/api/java/com/minecolonies/api/tileentities/TileEntityStash.java  6
- src/main/java/com/minecolonies/coremod/network/messages/ChangeDeliveryPriorityMessage.java  1
- src/main/java/com/minecolonies/coremod/colony/buildings/views/EmptyView.java  1
- src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/Stash.java  1
- src/main/java/com/minecolonies/coremod/blocks/huts/BlockStash.java  5
         

Clones added
============
- src/main/java/com/minecolonies/coremod/blocks/huts/BlockStash.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Minecolonies/minecolonies/pullRequest?prid=5282331&bid=17293574)

Shirkit committed:
Formatting using CheckStyle plugin

reggaemuffin commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/main/java/com/minecolonies/coremod/network/messages/ChangeDeliveryPriorityMessage.java  1
- src/main/java/com/minecolonies/coremod/colony/buildings/views/EmptyView.java  1
- src/main/java/com/minecolonies/coremod/blocks/huts/BlockStash.java  5
- src/main/java/com/minecolonies/coremod/network/messages/ChangeDeliveryPriorityStateMessage.java  1
- src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/Stash.java  1
- src/api/java/com/minecolonies/api/tileentities/TileEntityStash.java  6
         

Clones added
============
- src/main/java/com/minecolonies/coremod/blocks/huts/BlockStash.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Minecolonies/minecolonies/pullRequest?prid=5282331&bid=17293574)

Raycoms committed:
Merge branch 'version/1.15' into version/1.15

reggaemuffin commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/main/java/com/minecolonies/coremod/network/messages/ChangeDeliveryPriorityStateMessage.java  1
- src/api/java/com/minecolonies/api/tileentities/TileEntityStash.java  6
- src/main/java/com/minecolonies/coremod/blocks/huts/BlockStash.java  5
- src/main/java/com/minecolonies/coremod/network/messages/ChangeDeliveryPriorityMessage.java  1
- src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/Stash.java  1
- src/main/java/com/minecolonies/coremod/colony/buildings/views/EmptyView.java  1
         

Clones added
============
- src/main/java/com/minecolonies/coremod/blocks/huts/BlockStash.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Minecolonies/minecolonies/pullRequest?prid=5282331&bid=17293574)

Raycoms commented on src/api/java/com/minecolonies/api/tileentities/TileEntityStash.java:
```diff
@@ -0,0 +1,74 @@
+                {
+                    iBuilding.alterPickUpPriority(MAX_PRIO);
+                }
+            } else
```
else on other line

Shirkit replied:
Sorry, I thought I had the auto-formatter fixed this. I'm fixing this manually.

Raycoms commented on src/api/java/com/minecolonies/api/tileentities/TileEntityStash.java:
```diff
@@ -0,0 +1,74 @@
+        IBuildingContainer building = getBuilding();
+        if (!isEmpty && building != null && !building.isPriorityStatic())
+        {
+            if (building instanceof IBuilding)
```
why do we need a different handling for IBuilding here? isn't it always an IBuilding or never?

Shirkit replied:
It should be always an instance of `IBuilding`, but as so far as I've looked, I haven't read the full project to see if it could be reset somehow, that's why I left the check there. Removing it.

Raycoms replied:
To be fair, I would replace the building != null check with that instanceof check.

Raycoms commented on src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/Stash.java:
```diff
@@ -0,0 +1,61 @@
+ * Class used to manage the stash building block.
+ */
+public class Stash extends AbstractBuilding
+{
```
I know this is a bit out of the scope of this PR. But maybe you're up to it. Maybe it makes sense to extract what Stash and PostBox need from AbstractBuilding into a separate abstract super class and have Stash and PostBox extend those.

So that we don't need the "!(building instanceof PostBox) && !(building instanceof Stash)" checks in a row, that might be cleaner.

Or at least an IRSComponent interface that both Stash and PostBox implement and we check only if !instanceof IRSComponent

Shirkit replied:
I thought about that as well, but I figured it was a bit out of scope of the PR. Also, it wouldn't need an extraction only from `AbsctractBuilding`, but the whole hierarchy, as there's a lot things there that isn't needed whatsoever. And I actually thought of, maybe, letting Stash and PostBox be also built along with an schematic? I think it's a whole project on it's on right there, and being new to the project (and to Minecraft modding), I'm not confident enough to pull drastic changes such as these.

But the idea of the interface is actually good. The IRSComponent you thought would be just an empty interface?

Raycoms replied:
Yes, I agree. Yes it would just be an empty interface.

Shirkit committed:
Fixed model size while on inventory

Was overflowing to next tile

Shirkit committed:
Merge branch 'version/1.15' of https://github.com/Shirkit/minecolonies into version/1.15

reggaemuffin commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/api/java/com/minecolonies/api/tileentities/TileEntityStash.java  6
- src/main/java/com/minecolonies/coremod/network/messages/ChangeDeliveryPriorityStateMessage.java  1
- src/main/java/com/minecolonies/coremod/network/messages/ChangeDeliveryPriorityMessage.java  1
- src/main/java/com/minecolonies/coremod/blocks/huts/BlockStash.java  5
- src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/Stash.java  1
- src/main/java/com/minecolonies/coremod/colony/buildings/views/EmptyView.java  1
         

Clones added
============
- src/main/java/com/minecolonies/coremod/blocks/huts/BlockStash.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Minecolonies/minecolonies/pullRequest?prid=5282331&bid=17293574)

Shirkit committed:
It's always IBuilding, so we can cast directly

And I always forget that we can cast null values without exceptions

reggaemuffin commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/main/java/com/minecolonies/coremod/network/messages/ChangeDeliveryPriorityMessage.java  1
- src/main/java/com/minecolonies/coremod/blocks/huts/BlockStash.java  5
- src/api/java/com/minecolonies/api/tileentities/TileEntityStash.java  5
- src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/Stash.java  1
- src/main/java/com/minecolonies/coremod/network/messages/ChangeDeliveryPriorityStateMessage.java  1
- src/main/java/com/minecolonies/coremod/colony/buildings/views/EmptyView.java  1
         

Clones added
============
- src/main/java/com/minecolonies/coremod/blocks/huts/BlockStash.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Minecolonies/minecolonies/pullRequest?prid=5282331&bid=17293574)

Shirkit committed:
Proper check and casting

reggaemuffin commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/main/java/com/minecolonies/coremod/network/messages/ChangeDeliveryPriorityMessage.java  1
- src/main/java/com/minecolonies/coremod/blocks/huts/BlockStash.java  5
- src/main/java/com/minecolonies/coremod/network/messages/ChangeDeliveryPriorityStateMessage.java  1
- src/api/java/com/minecolonies/api/tileentities/TileEntityStash.java  5
- src/main/java/com/minecolonies/coremod/colony/buildings/views/EmptyView.java  1
- src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/Stash.java  1
         

Clones added
============
- src/main/java/com/minecolonies/coremod/blocks/huts/BlockStash.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Minecolonies/minecolonies/pullRequest?prid=5282331&bid=17293574)

Shirkit committed:
Implemented IRSComponent

reggaemuffin commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/main/java/com/minecolonies/coremod/colony/buildings/views/EmptyView.java  1
- src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/Stash.java  1
- src/main/java/com/minecolonies/coremod/blocks/huts/BlockStash.java  5
- src/main/java/com/minecolonies/coremod/network/messages/ChangeDeliveryPriorityStateMessage.java  1
- src/api/java/com/minecolonies/api/tileentities/TileEntityStash.java  5
- src/main/java/com/minecolonies/coremod/network/messages/ChangeDeliveryPriorityMessage.java  1
         

Clones added
============
- src/main/java/com/minecolonies/coremod/blocks/huts/BlockStash.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Minecolonies/minecolonies/pullRequest?prid=5282331&bid=17293574)

Shirkit commented:
So @Raycoms, unfortunately, the check I did for `IRSComponent` is not preventing the pop-up asking for the user to use the Build Tool to show up on the case of Stash.

I'll take a closer look why that is the case. `BlockPostBox` didn't result in anything meaningful, just the `PostBox`, which I've implemented `IRCSComponent`. I'm digging up here, but haven't figure out why that is the case.

Raycoms commented:
EventHandler.java: if (block instanceof AbstractBlockHut && block != ModBlocks.blockPostBox)
could use IRSComponentBlock as an additional interface here.

Shirkit committed:
Added empty interface for non-schematic buildings

Shirkit commented:
I think the PR is ready

reggaemuffin commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/main/java/com/minecolonies/coremod/blocks/huts/BlockStash.java  5
- src/main/java/com/minecolonies/coremod/network/messages/ChangeDeliveryPriorityStateMessage.java  1
- src/main/java/com/minecolonies/coremod/colony/buildings/views/EmptyView.java  1
- src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/Stash.java  1
- src/api/java/com/minecolonies/api/tileentities/TileEntityStash.java  5
- src/main/java/com/minecolonies/coremod/network/messages/ChangeDeliveryPriorityMessage.java  1
         

Clones added
============
- src/main/java/com/minecolonies/coremod/blocks/huts/BlockStash.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Minecolonies/minecolonies/pullRequest?prid=5282331&bid=17293574)

Raycoms committed:
Merge branch 'version/1.15' into version/1.15

marvin-bitterlich commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/Stash.java  1
- src/main/java/com/minecolonies/coremod/blocks/huts/BlockStash.java  5
- src/api/java/com/minecolonies/api/tileentities/TileEntityStash.java  5
- src/main/java/com/minecolonies/coremod/colony/buildings/views/EmptyView.java  1
- src/main/java/com/minecolonies/coremod/network/messages/ChangeDeliveryPriorityStateMessage.java  1
- src/main/java/com/minecolonies/coremod/network/messages/ChangeDeliveryPriorityMessage.java  1
         

Clones added
============
- src/main/java/com/minecolonies/coremod/blocks/huts/BlockStash.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/Minecolonies/minecolonies/pullRequest?prid=5282331&bid=17293574)

marchermans reviewed (changes requested):
Very good work. Some minor nitpicking. You made interfaces to mark classes as an RSComponeny. (Both in block and building variant). However it would be amazing if you could give the logic that handles the deliveryman priorities the same treatment. Reducing the instanceof abstractbuilding or instanceof stash checks and providing a better logic for the future.

Would you be willing to add it?

Shirkit commented:
> Very good work. Some minor nitpicking. You made interfaces to mark classes as an RSComponeny. (Both in block and building variant). However it would be amazing if you could give the logic that handles the deliveryman priorities the same treatment. Reducing the instanceof abstractbuilding or instanceof stash checks and providing a better logic for the future.
> 
> Would you be willing to add it?

That would require extracting the priority part from the `AbstractBuildingContainer` into a separate interface, which I think it's a bit out of scope from this PR, as I told @Raycoms. Not familiar enough with the project, and by taking a look, I prefer not to pick such work after dropping from parachute into Java again after 4 years.

Raycoms commented:
no problem, can be left to a future task for sure. If you solve the small conflicts I can merge this PR then =)

marchermans reviewed (dismissed):
Fix the merge conflicts and you are good to go.

Shirkit committed:
Merge branch 'version/1.15' into version/1.15

Shirkit commented:
I think it should be all good now.
